pkg/apis/acme/v1alpha2/types_order.go

```diff
@@ -168,15 +168,31 @@ type ACMEChallenge struct {
 }
 
 // ACMEChallengeType denotes a type of ACME challenge
-// +kubebuilder:validation:Enum=http-01;dns-01
+// +kubebuilder:validation:Enum=http-01;dns-01;tls-alpn-01;tls-sni-01;tls-sni-02
```

Comment:
i don't think we should validate these.

The source of these comes from the API not the user, the use case for validation is minimal. But that doesn't fully make the case.

`"tls-alpn-01"` already makes a point here that the spec allows to add more validation methods. cert-manager misbehaves is it find a challenge type that isn't in this list as the API is not able to store it as it is "invalid", this causes all old cert-manager versions to not be able to issue certificates the moment let's encrypt (or another ACME issuer) adds any new validation type

Reply:
i found this out the hard way in https://github.com/jetstack/cert-manager/pull/2704/commits/799d2535d0e5ef8d2695674916d7b82535b40b09

Reply:
I am not so sure - our own controllers only support a limited set of types, so I was considering also modifying the controller as part of this PR to reject unknown challenge types too.

Whilst `tls-alpn-01` does demonstrate new ones can be added, we don't have an extensibility story that will actually allow these to be implemented just yet. I feel like we'd be better to leave the allowed values as `http-01` and `dns-01` *only* and then throw anything that isn't recognised away? Not sure 🤷‍♂️ 

If, for example, a user creates a Challenge resource manually and specifies `type` as `NOT_REAL`, it'd be far more useful if that create call fails and says "only HTTP-01 and DNS-01 allowed"

WDYT?

Reply:
I am pro limiting these to the ones we support where our controller is fully tested for that behavior.